src/nvrpc/BidirectionalStreamingLifeCycle.h

```diff
@@ -0,0 +1,366 @@
+  static ServiceQueueFuncType BindServiceQueueFunc(
+      /*
+      std::function<void(
+          ServiceType *, ::grpc::ServerContext *,
+          ::grpc::ServerAsyncReaderWriter<ResponseType, RequestType>*,
+          ::grpc::CompletionQueue *, ::grpc::ServerCompletionQueue *, void *)>
+      */
+      RequestFuncType request_fn, ServiceType* service_type)
+  {
+    return std::bind(
+        request_fn, service_type,
+        std::placeholders::_1,  // ServerContext*
+        std::placeholders::_2,  // AsyncReaderWriter<OutputType, InputType>
+        std::placeholders::_3,  // CQ
+        std::placeholders::_4,  // ServerCQ
+        std::placeholders::_5   // Tag
+    );
+  }
+
+  static ExecutorQueueFuncType BindExecutorQueueFunc(
+      ServiceQueueFuncType service_q_fn, ::grpc::ServerCompletionQueue* cq)
+  {
+    return std::bind(
+        service_q_fn,
+        std::placeholders::_1,  // ServerContext*
+        std::placeholders::_2,  // AsyncReaderWriter<Response, Request> *
+        cq, cq,
+        std::placeholders::_3  // Tag
+    );
+  }
+};
+
+// Implementation
+template <class Request, class Response>
+bool
+BidirectionalStreamingLifeCycle<Request, Response>::RunNextState(bool ok)
+{
+  return (this->*m_NextState)(ok);
+}
+
+
+template <class Request, class Response>
+BidirectionalStreamingLifeCycle<
+    Request, Response>::BidirectionalStreamingLifeCycle()
+    : m_ReadStateContext(static_cast<IContext*>(this)),
+      m_WriteStateContext(static_cast<IContext*>(this))
+{
+  m_ReadStateContext.m_NextState = &BidirectionalStreamingLifeCycle<
+      RequestType, ResponseType>::StateRequestDone;
+  m_WriteStateContext.m_NextState = &BidirectionalStreamingLifeCycle<
+      RequestType, ResponseType>::StateResponseDone;
+}
+
+template <class Request, class Response>
+void
+BidirectionalStreamingLifeCycle<Request, Response>::Reset()
+{
+  std::queue<RequestType> empty_request_queue;
+  std::queue<ResponseType> empty_response_queue;
+  std::queue<ResponseType> empty_response_write_back_queue;
+  OnLifeCycleReset();
+  {
+    std::unique_lock<std::mutex> lock(m_QueueMutex);
+    m_Executing = false;
+    m_RequestQueue.swap(empty_request_queue);
+    m_ResponseQueue.swap(empty_response_queue);
+    m_ResponseWriteBackQueue.swap(empty_response_write_back_queue);
+    m_Context.reset(new ::grpc::ServerContext);
+    m_ReaderWriter.reset(
+        new ::grpc::ServerAsyncReaderWriter<ResponseType, RequestType>(
+            m_Context.get()));
+  }
+  m_NextState = &BidirectionalStreamingLifeCycle<
+      RequestType, ResponseType>::StateInitializedDone;
+  m_QueuingFunc(m_Context.get(), m_ReaderWriter.get(), IContext::Tag());
+}
+
+// The following are a set of functions used as function pointers
+// to keep track of the state of the context.
+template <class Request, class Response>
+bool
+BidirectionalStreamingLifeCycle<Request, Response>::StateInitializedDone(
+    bool ok)
+{
+  if (!ok)
```

Comment:
probably should scope this with {}.  my bad as the reference didn't do it either.